### ISSUE TITLE
fix(metrics/loc): sloc is 0 for a one-line unterminated file

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1067]
## Summary

`Sloc::sloc()` computes `end - start` with no `+ 1` for unit spans, so a
one-line file reports `sloc == 0`. That feeds `mi::inputs_are_empty`, which
then returns `0.0` for a real, non-empty file.

Found during the #1051 review; pre-existing, affects every language.

## Detail

The unit branch is only correct because a trailing newline gives the root
`source_file` a phantom final row. `Source` / `analyze` apply no
normalisation (`src/spaces/ast.rs:34-45`), so any caller passing bytes
directly hits the off-by-one — including the documented rustdoc example at
`src/spaces.rs:313`:

```rust
analyze(Source::new(LANG::Rust, b"fn main() {}"), MetricsOptions::default())
```

yields `sloc() == 0` for a one-line file. `mi::inputs_are_empty` (`mi.rs:62`)
then sees `sloc <= 0` and returns `0.0` from `original()` / `sei()` /
`visual_studio()`.

There is a second symptom on the same input class: for `b"fn f(){}\n/// x"`,
`sloc() == 1` while `ploc() == 1` and `cloc() == 1`, so `cloc + ploc > sloc`
— violating the invariant documented at `src/metrics/loc.rs:448` ("`cloc <=
sloc` always holds for parsed input after #461").

## Why it was not caught

Same reason as #1051: every in-tree harness normalises. `check_func_space`
trims and re-appends a trailing newline, and the integration suites route
through `read_file_with_eol`, whose `normalize_line_endings` ends with an
unconditional `data.push(b'\n')`. The input class is unreachable from either.
`src/tools.rs`'s `metrics_verbatim` is the helper to use for a regression
test.

## Caution

This changes metric values for un-newline-terminated input, so it needs the
same treatment #1051 got: state the drift in the CHANGELOG explicitly and
check whether any baseline or snapshot moves. Fixing `sloc` may also shift MI
for the affected files.

---

## Resolution

**Status**: Fixed (pending merge)
**Commit**: cec20358 on branch `fix/issue-1067`
**Root cause**: `Sloc::sloc()` selected its row count from an
`is_unit` flag rather than from the span's end column. The unit
branch (`end - start`) is right only when a trailing newline pushes
the root node's end onto a phantom row, and the non-unit branch
(`end - start + 1`) is wrong wherever a grammar node genuinely ends
at column 0 — which tree-sitter-perl's `function_definition` does
after the last `sub` in a file. Both branches collapse into one
`span_rows(start_row, end_row, end_column)` rule. The reported
`cloc + ploc > sloc` violation is the same off-by-one, not a
separate defect; the MI zeroing falls out once `sloc > 0`.

**Comment by dekobon** [comment]
## Fix Summary

**Root cause**: `Sloc::sloc()` chose its row count from an `is_unit`
flag — `end - start` for the unit span, `end - start + 1` for every
other space. Both are proxies for the actual question: *does the
span's end position sit at column 0?* If it does, that final row
contributes no characters and must not be counted; if it does not,
it must.

The proxy was wrong in both directions:

- tree-sitter's root node runs to end-of-input, so the unit ends at
  column 0 only when the file is newline-terminated. Source that
  stops mid-line lost its last row — hence `sloc == 0` for
  `b"fn main() {}"`, `mi::inputs_are_empty` short-circuiting MI to
  `0.0`, and `cloc + ploc > sloc` for `b"fn f(){}\n/// x"`. All three
  are the same off-by-one, so the second symptom is not a separate
  defect.
- The unconditional `+ 1` for non-unit spaces was wrong wherever a
  grammar node *does* end at column 0. A debug-assert sweep over the
  whole test suite found exactly one such case: tree-sitter-perl's
  `function_definition` swallows the newline after the closing brace
  of a file's **last** `sub`, so that sub's `sloc` was inflated by one
  — reporting a 3-row `sub` as 4, and (in
  `perl_anonymous_sub_in_sub_lloc`) a `sloc_max` of 5 for a 4-row file.

**Changes**:
- `src/metrics/loc.rs`: new `span_rows(start_row, end_row, end_column)`
  helper; `Sloc::unit: bool` replaced by `Sloc::end_column: usize`;
  `sloc()` and `exclude_span()` both go through `span_rows`, so the
  `exclude_tests` prune accounting follows the same rule as the span
  it subtracts from. `Loc::compute` drops its `is_unit` parameter.
- `src/metrics/loc/shared.rs`: `init` reads `node.end_position()` and
  records the end column instead of the unit flag.
- The 23 `src/metrics/loc/<lang>.rs` modules and
  `src/macros/mod.rs`: mechanical `is_unit` removal.
- `src/spaces/compute.rs`: `NodeFacts::unit` removed; with the Loc
  consumer gone, `get_space_kind_with_code` is now called only for
  nodes that open a space (the #522 always-on lookup is no longer
  needed). `init_unit_span` / `exclude_test_span` take the end column.
- `src/spaces.rs`: the `Source` rustdoc example cited in the issue now
  asserts `sloc() == 1` — it previously asserted only the space name,
  so nothing there was *wrong*, but the fixed behaviour is worth
  pinning where the bug was documented.
- `src/test_support.rs`: new `space_verbatim` (byte-for-byte analysis
  returning the whole `FuncSpace`) so per-function assertions are
  possible without normalisation; `metrics_verbatim` delegates to it.

**Languages affected**: all of them for the unit-span half (`sloc` is
computed once in the shared `Sloc`); Perl additionally for the
per-function half.

**Metric drift** (user-visible):
- Library callers passing un-newline-terminated bytes through
  `Source` / `Ast::parse` now get the real row count: `sloc` +1 (and
  `blank`, `sloc_max`, the `*_average` values, and all three MI
  formulas move with it — MI was previously pinned to `0.0` for
  one-line files). The CLI, web server and Python bindings all read
  through `read_file_with_eol`, which appends a trailing newline, so
  none of them change on this axis.
- Every Perl file's last `sub` loses one `sloc` row, plus the file's
  `sloc_max` / `blank_max` where that sub was the maximum. This one
  *does* reach the CLI/web/Python.
- Whitespace-only files with no trailing newline go from `sloc 0` to
  `sloc 1` / `blank 1`.

No public API break: `end_column` is private, and `Loc`,
`init_unit_span`, `exclude_test_span` are all crate-internal.
`.bca-baseline.toml` did not move (no Perl in the repo, and every
source file is newline-terminated).

**Tests** (all in `src/metrics/loc.rs`, all via `metrics_verbatim` /
`space_verbatim` — a `check_metrics` test would pass vacuously since
that helper re-appends a trailing newline):
- `unterminated_one_line_file_reports_one_source_line` — a one-line
  unterminated fixture for **every** language whose `Loc` impl is not
  a documented no-op (23 entries; `Preproc` / `Ccomment` excluded per
  #188). Asserts `sloc == 1`, `sloc_max == 1`, and
  `cloc + ploc <= sloc`.
- `trailing_newline_does_not_change_loc_or_mi` — same fixtures, with
  and without a trailing newline: all five LOC sub-metrics and all
  three MI formulas must agree, and MI must not be zero. This pins
  the newline-terminated path too, so the fix cannot silently move
  the case that was already correct.
- `unterminated_trailing_comment_upholds_the_cloc_ploc_invariant` —
  the `b"fn f(){}\n/// x"` case from the issue.
- `degenerate_inputs_report_their_real_row_count` — empty input,
  whitespace-only unterminated, and the (unchanged) whitespace-only
  newline-terminated case where tree-sitter collapses the root to a
  zero-width node and there is no span left to read.
- `perl_last_sub_does_not_absorb_the_trailing_newline` — two 3-row
  subs, both must report 3.
- Two existing Perl snapshots updated (`sloc_max` 5→4, `blank_max`
  1→0 — the old values exceeded the 4-row file's own `sloc`).

Verified by test-via-revert per `.claude/rules/testing.md`: with
`span_rows` perturbed back to `end - start`, all five new tests fail;
perturbed to `end - start + 1`, the Perl and degenerate-input tests
fail.

Integration-snapshot submodule: **no drift** — the corpus has no Perl
and every file is newline-terminated, so
`tests/repositories/big-code-analysis-output` is untouched.

`make pre-commit` passes clean.

**Commit**: cec20358 on branch fix/issue-1067
